[Form] Added CountryType option for using alpha3 country codes

creiner opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no 
| Tickets       | Fixes #20313
| License       | MIT
| Doc PR        | -

In the linked issue #20313 was a proposal to add an alpha3 option to the country type.
Here it is..
Hopefully I've made no mistake, so when the code is fine, I will create a documentation PR.. :-)

creiner committed:
[Form] Added CountryType option for using alpha3 country codes

ro0NL commented on src/Symfony/Component/Form/Extension/Core/Type/CountryType.php:
```diff
@@ -42,13 +42,15 @@ public function configureOptions(OptionsResolver $resolver)
             },
             'choice_translation_domain' => false,
             'choice_translation_locale' => null,
+            'alpha3' => false,
```
you should set the allowed types (bool) below as well

creiner replied:
Good idea.. i've added it..

ro0NL commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/CountryTypeTest.php:
```diff
@@ -58,6 +58,25 @@ public function testChoiceTranslationLocaleOption()
     }
 
+    /**
+     * @requires extension intl
```
i believe this is only needed with `choice_translation_locale` which triggers a different collation for sorting

creiner replied:
Aaah good to know.. I've removed it..

ro0NL commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/CountryTypeTest.php:
```diff
@@ -58,6 +58,25 @@ public function testChoiceTranslationLocaleOption()
+    /**
+     * @requires extension intl
+     */
+    public function testAlpha3Option()
```
perhaps we should add a 2nd test (testAlpha3OptionWithTranslationLocale) to ensure alpha3 codes are translatable as well. See inspiration above.

creiner replied:
I've added it!

ro0NL reviewed (commented):
the same applies to LanguageType, isnt it?

creiner commented:
I've added it to the LanguageType and added also the tests...

When a language doesn't have a three letter code, it doesn't appear in the collection (behavior of Language::getAlpha3Names ).. 
Maybe this can cause confusion.. do you think a fallback to 2-letter codes for languages which doesn't have three letter codes would be appropriate?

ro0NL reviewed (commented):
LGTM. I think you need to bump symfony/intl in composer.json of the Form component still ... see https://travis-ci.org/symfony/symfony/jobs/592410645#L4077

> do you think a fallback to 2-letter codes for languages which doesn't have three letter codes would be appropriate?

i'd like to validate the dataset further in #33165, feel free to share any data related issues there. In general the alpha2 and alpha3 list should be the same size. For languages the alpha2 lists also includes alpha3 codes in case no alpha2 code is available, whereas Countries should have alpha3 codes for each alpha2 code.

fabpot commented:
Thank you @creiner.
